frame/dapps-staking/src/pallet/mod.rs

```diff
@@ -978,6 +1016,81 @@ pub mod pallet {
 
             Ok(().into())
         }
+
+        // TODO: benchmark this call
+        /// Used to set reward delegate for staker rewards.
+        /// 
+        /// The caller must be the staker or the current delegate.
+        /// Reward delegation will only apply when RewardDestination of the staker is set to free balance.
+        /// 
+        /// - `staker`: account id of the staker. 
+        /// - `contract_id`: smart contract id.
+        /// - `new_delegate`: account id of the new delegate.
+        #[pallet::weight(0)]
+        pub fn set_reward_delegate(
+            origin: OriginFor<T>,
+            staker: T::AccountId, // the caller must be the staker or the current delegate
+            contract_id: T::SmartContract,
+            new_delegate: T::AccountId,
+        ) -> DispatchResult {
+            Self::ensure_pallet_enabled()?;
+            let caller = ensure_signed(origin)?;
+
+            let current_delegate = RewardDelegate::<T>::get(&staker, &contract_id);
+
+            // Check the caller is the staker or the current delegate
+            ensure!(
+                caller == staker || current_delegate == Some(caller),
+                Error::<T>::NotAuthorizedToDelegate
+            );
+
+            // Ensure the staker does not have reward destination set to StakeBalance
+            let ledger = Self::ledger(&staker);
+            ensure!(
+                ledger.reward_destination != RewardDestination::StakeBalance,
+                Error::<T>::RewardDestinationStakeBalance
+            );
```

Comment:
It will make no difference, because restaking and delegation now depends on delegate account being set or not. I though about creating a new reward destination so to make it more clear, but as it is implemented now, the rewards destinations are associated to the ledger, so are associated to an account. And for the delegation logic something smart contract specific was needed. I still think that is a possibility worth exploring.